crates/future_metrics/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+    },
+};
+
+/// Target specified in [`metrics::Metadata`] for all metrics produced by this
+/// crate.
+pub const METADATA_TARGET: &str = "future_metrics";
+
+/// Metric names used by this crate.
+pub mod metric_name {
+    pub const FUTURE_DURATION: &str = "future_duration";
+
+    pub const FUTURES_CREATED: &str = "futures_created";
+    pub const FUTURES_STARTED: &str = "futures_started";
+    pub const FUTURES_FINISHED: &str = "futures_finished";
+    pub const FUTURES_CANCELLED: &str = "futures_cancelled";
+
+    pub const FUTURE_POLL_DURATION: &str = "future_poll_duration";
+    pub const FUTURE_POLL_DURATION_TOTAL: &str = "future_poll_duration_total";
+}
+
+/// Creates a new label identifying a future by its name.
+pub const fn future_name(s: &'static str) -> Label {
+    Label::from_static_parts("future_name", s)
+}
+
+pub trait FutureExt: Sized {
+    /// Consumes the future, returning a new future that records the executiion
+    /// metrics of the inner future.
+    ///
+    /// It is expected that you provide at least one label identifying the
+    /// future being metered.
+    /// Consider using [`future_name`] label, or the [`FutureExt::with_metrics`]
+    /// shortcut.
+    fn with_labeled_metrics(self, labels: &'static [Label]) -> Metered<Self> {
+        Metered::new(self, labels)
+    }
+
+    /// A shortcut for [`FutureExt::with_labeled_metrics`] using a single label
+    /// only (presumably [`future_name`]).
+    fn with_metrics(self, label: &'static Label) -> Metered<Self> {
+        self.with_labeled_metrics(std::slice::from_ref(label))
+    }
+}
+
+impl<F> FutureExt for F where F: Future {}
+
+#[pin_project::pin_project]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Metered<F> {
+    #[pin]
+    future: F,
+    state: State,
+}
+
+struct State {
+    started_at: Option<Instant>,
+    is_finished: bool,
+
+    poll_duration_total: Duration,
+
+    metrics: Metrics,
+}
+
+impl<F> Metered<F> {
+    fn new(future: F, metric_labels: &'static [Label]) -> Self {
+        let metrics = Metrics::new(metric_labels);
+
+        metrics.created.increment(1);
+
+        Self {
+            future,
+            state: State {
+                started_at: None,
+                is_finished: false,
+                poll_duration_total: Duration::from_secs(0),
+                metrics,
+            },
+        }
+    }
+}
+
+impl<F: Future> Future for Metered<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let s = &mut this.state;
+
+        if s.started_at.is_none() {
+            s.started_at = Some(Instant::now());
+            s.metrics.started.increment(1);
+        }
+
+        let poll_started_at = Instant::now();
+        let result = this.future.poll(cx);
+        let poll_duration = poll_started_at.elapsed();
+
+        s.metrics.poll_duration.record(poll_duration);
```

Comment:
You should definitely not `record()` metrics on each poll entry, that's way too expensive. At least that was the case with `opentelemetry`, but I believe it would still be expensive with other backends too. I suggest storing these values as `Duration` on the future, and `record()` on `drop()`. In fact, you're already adding this duration to `s.poll_duration_total`, so may as well just move this `record()` to `drop()`.

Reply:
The idea is to be able to detect incorrectly implemented futures - the ones using blocking code, or the ones that take too much time in a single poll.
By moving it to `drop` we would lose the ability to use histogram buckets. Total duration is already being calculated, so it's going to be the same metric. 

I'll make this metric optional for now then, at least until we know the performance impact

Reply:
It's hard to say where's more value - in understanding the total CPU time spent in the poll, or a single poll duration. We could do both by adding an internal poll entries counter, and then on `drop()` we'd record two metrics: average poll duration and total poll duration. Both could be histograms. Obviously it's not as precise if you use poll entry count to calculate the average, but it would be light on performance and would give us just enough of insight into the future.

I would also suggest adding the `poll_entries` metrics. It was helpful to me a few times.

Reply:
we can already get the average by just using the `poll_duration_total`, because it contains the `count` of records

Reply:
> I would also suggest adding the poll_entries metrics. It was helpful to me a few times.
`poll_duration_total` produces this metric indirectly as it's a histogram

Reply:
I don't get it. How's that possible if you're only recording it once on `drop()`? I meant the number of poll entries per future executed. Not the total number of futures polled. It's possible to get this value from `poll_duration`, but if you're removing/reworking it, you'd need a separate counter for that.

Reply:
Ah, sorry, yes, I was talking with assumption that `poll_duration` is still there